### DISCUSSION TITLE
FastAPI public solve_dependencies API

### DISCUSSION THREAD

**Discussion by Lancetnik** [fastapi/fastapi#12131] [Show and tell]
Hello there! @tiangolo as you asked in Discord, I'll try to describe my case here

I am a core maintainer of [**FastStream**](https://github.com/airtai/faststream) project. As one of the feature we provide - we have pretty close **FastAPI** integration ([documentation](https://faststream.airt.ai/latest/getting-started/integrations/fastapi/))

In a few words: using such plugin you can register **RMQ**/**Kafka**/**Redis**/**NATS**/**SQS**/etc subscribers the same way with regular **FastAPI** HTTP endpoints

```python
from fastapi import FastAPI
from pydantic import BaseModel

from faststream.rabbit.fastapi import RabbitRouter

router = RabbitRouter("amqp://guest:guest@localhost:5672/")

class Incoming(BaseModel):
    m: str

@router.subscriber("in-queue")
@router.publisher("out-queue")
async def hello(m: Incoming):
    return {"response": "Hello, Rabbit!"}

app = FastAPI()
app.include_router(router)
```

The main feature of this integration is full **FastAPI** serialization and **Depends** system support - you can use regular `fastapi.Depends/Path/Header/BackgroundTasks/etc` options to serialize your incloming message and reuse already written **FastAPI** dependencies.

It is a pretty simple and limited code, but we have to use **FastAPI** "private" methods like `get_dependent`, `solve_dependencies` and other.

The main mechanism really simple. We register regular **FastStream** subscriber, but wrap the original function to **FastAPI**-compatible decorator before. So, it is working like this: **FastStream** subscriber -> **FastAPI** wrapper -> user' handler

As you can see, [this wrapper](https://github.com/airtai/faststream/blob/main/faststream/broker/fastapi/route.py#L176) is just a some kind of copy of original **FastAPI** [endpoint function](https://github.com/fastapi/fastapi/blob/master/fastapi/routing.py#L290).

But, this function is private and too tied to HTTP logic. I think, it will be very helpful, if **FastAPI** has some public API for `get_dependant` and `run_endpoint` (incapsulates `solve_dependencies`, `run_endpoint_function` and `serialize_response` functions) with stable signature. It will be helpful for me (and other future plugin developers) to reuse **FastAPI** serialization logic as separated functional, and for **FastAPI** itself to better organize code layers and responsibility as well. Also, I can help with such refactoring if this feature will be approved.

**Comment by Tishka17** [comment]
It would be really nice to have a place where to put your own dependency solver

**Comment by Lancetnik** [comment]
@tiangolo sorry for ping, but you asked to create such discussion. Could you spend some of your attention here?

**Comment by tiangolo** [comment]
Hello! @YuriiMotov made me aware that the latest release had incompatible changes with FastStream and pointed me here.

I came to read and try to understand the use case better, but I'm struggling to see what the benefit of having a FastAPI integration would be. :thinking: 

The FastAPI dependency system is super simple, and coupled with the HTTP request data, as you have noted.

And I see you have your own internal `Depends` in FastStream. I would think it would be better and simpler to tell people something like:

> You can use many of the same dependency functions with FastStream, as long as they don't require anything from the request, for example:

```Python
import fastapi
import faststream

def some_dependency() -> int:
    return 1

FastAPIDep = Annotated[int, fastapi.Depends(some_dependency)]
FastStreamDep = Annotated[int, faststream.Depends(some_dependency)]
```

> And then you could use the `FastAPIDep` in FastAPI functions and `FastStreamDep` in FastAPI functions.

Or alternatively, you could also support:

```Python
FastAPIDep = Annotated[int, fastapi.Depends(some_dependency), faststream.Depends(some_dependency)]
```

---

The thing is, FastAPI dependency functions are normally defined in a way that is quite tightly coupled with FastAPI HTTP stuff, so I would imagine they wouldn't be able to just re-use the same dependency functions. And if they do, then those are probably very simple dependency functions (e.g. to get a DB session), and they can re-use the function with the new dependency as in the example above.

I struggle to see the benefit of bending how FastAPI works to solve dependencies for something that is not related to FastAPI or HTTP in general, and that also has its own equivalent way to achieve the same. :thinking: 

Could you share some examples of why people need this?

**Comment by Lancetnik** [comment]
@tiangolo Hi! Thank you for your feedback. It's really valuable and I agree with you that this solution should work. I have the same opinion as you, but not all users. About 50% of users use our integration instead of native FastAPI + FastStream applications together. The main advantage of this integration is the ability to seamlessly use FastAPI's serialization, DI and BackgroundTasks mechanisms in endpoints to consume messages from RabbitMQ/Kafka/etc. In addition, it adds an AsyncAPI documentation section for these endpoints to the main application. You can check all features in the doc - https://faststream.ag2.ai/latest/getting-started/integrations/fastapi/

Since the current FastAPI syntax encourages people to create dependencies separately using the Annotated `Dep = Annotated[type, fastapi.Depends()]` mechanism, adding a FastStream endpoint to the same application for message consumption does not allow for native reuse of previously declared dependencies. Instead, you'll need to go through each one and change it by adding Dep = Annoated[type, fastapi.Depens(), faststream.Depends()]`. Also, although faststream depends are a copy of fastapi depends, the behavior may differ in certain details (for example, I haven't yet implemented or am unsure about whether I will add the scope feature from 0.121). Therefore, it would be more convenient and predictable for users if they could reuse all their previous FastAPI experience and existing Depends when working with event-driven endpoints, as if they were using the same framework.

For example, the Nest framework from NodeJS allows users to declare endpoints for any protocol - HTTP/AMQP/Kafka/MQTT/etc. - and the user loves it. I believe that this kind of integration with FastAPI can significantly enrich its ecosystem. Thus, it could be not just an "HTTP microframework", but the core of an entire ecosystem of ready-made modules where you can simply `pip install AMQP / scheduler / authorization / S3 / etc` and reuse all your experience from FastAPI without the need for additional learning of tools.

I want to apologize for being so intrusive earlier. I understand how difficult it is to maintain such a large framework. I admire your work and just cannot imagine how I could support my tool if overwhelmed by such flood of demands and requests. I do not want to be intrusive, but I am inspired by FastAPI and want its ecosystem to be richer. However, we need some coordination for this.

**Reply by tiangolo** [reply]
Hi @Lancetnik, thanks for collecting all the ideas here!

Let me answer everything in line:

> The main advantage of this integration is the ability to seamlessly use FastAPI's serialization, DI and BackgroundTasks mechanisms in endpoints to consume messages from RabbitMQ/Kafka/etc. In addition, it adds an AsyncAPI documentation section for these endpoints to the main application. You can check all features in the doc - https://faststream.ag2.ai/latest/getting-started/integrations/fastapi/

Let me split this into subsections to answer each.

> FastAPI's serialization

I would imagine you already have good serialization in FastStream, I would think it's better to keep that than to use FastAPI for this.

> BackgroundTasks

If users are already invested in a system that depends on Kafka or similar, message passing/events, each function handler is already sort of a background task, I wouldn't think the integrated lightweight `BackgroundTasks` would be too useful here.

In fact, as they are intended only for minor things that could never be run, if the app instance is shutdown before they are finished, I would think it could be problematic to use it. I would advise against it, and would guide users to *not* do that.

> AsyncAPI

I have considered if there would make any sense to have some direct integration for AsyncAPI in FastAPI, but I haven't found a use case that justifies it. I also haven't researched it thoroughly.

For now, my current perspective would be that it would be better to have the separated apps, FastAPI with its own OpenAPI and the FastStream app with its own independent AsyncAPI.

From my current point of view, it wouldn't make sense to mix both, as they are normally different separated things/components.

> DI

I think the main point is the dependency injection.

The FastAPI dependency injection system is very simple, and coupled with extracting data from the request, that's one of the main reasons for it to exist. If there's no request, I don't think it would be that useful.

As you already have a dependency injection system, I would think it would be better to use it instead of having another mode, using FastAPI internals for it.

> Instead, you'll need to go through each one and change it by adding `Dep = Annoated[type, fastapi.Depens(), faststream.Depends()]`

If users have a dependency that is simple enough to be compatible with both FastAPI and FastStream it would mean it doesn't depend on request-specific data, so, for example it would be just for getting a DB session.

In those cases, there probably wouldn't be a long dependency chain. And if there was, for example, to get the `CurrentAdminUser`, that would still have to get some info from the request, e.g. authentication credentials for the user, which would make that part not really compatible with something that doesn't have a request.

I personally would think it would be better to have separated dependencies for both types of apps. But if users really, really wanted to have them mixed, e.g. they provide a very simple feature, as a DB session, then I would think it would be better for them to declare both there.

> Also, although faststream depends are a copy of fastapi depends, the behavior may differ in certain details (for example, I haven't yet implemented or am unsure about whether I will add the scope feature from 0.121). Therefore, it would be more convenient and predictable for users if they could reuse all their previous FastAPI experience and existing Depends when working with event-driven endpoints, as if they were using the same framework.

I think this reinforces the previous point, it probably makes sense to keep FastStream dependency injection system separated from FastAPI's, as the one in FastAPI is not designed for FastStream and might not accommodate its needs.

> About 50% of users use our integration instead of native FastAPI + FastStream applications together.

I don't think it would be a bad thing to guide users to use what you consider the correct way. I would think it would be better to teach them to use the same FastStream components to define dependencies instead of relying on the one in FastAPI, that are not designed for FastStream.

For example, in the past, I have rejected feature requests because I don't consider them appropriate, even if there are multiple users saying they want that feature.

On the other side, I have also accepted feature requests that I didn't fully agree with, only because several people were asking for it, and then I ended up regretting it, as they ended up bringing more problems down the line than what they helped.

So, I think it's okay to guide users to use things the way you intended even if some are asking for something else, as they probably don't have all the context you already have to help you know what are the best ways to handle a specific feature.

## Feature request - external dependency injection system 

From all this, what I think would seem to be the main request would be a dependency injection system independent of FastAPI that can be reused by other tools and tweaked by each, e.g. in FastAPI to extract the data from the request.

This is something I have considered for a while, but I haven't found the right use cases to justify it, as it would also need to be done in a way that keeps the current simplicity in FastAPI. And I haven't put the time into researching it yet.

I wouldn't discard this idea, but it's something I sadly can't prioritize right now. There are several fixes and features I need to handle in FastAPI first.

## FastAPI plans

In FastAPI, I consider internals the things that are not documented. At some point I intend to also refactor them to make them more explicitly private (prefixed with `_`), and then see what are the things I could need to expose externally, this would probably happen in consecutive releases.

This would involve renaming internal components, which would probably affect external tools that touch those internals. With this I will be able to see what things "break" to then see what should be the right way to do them, and have anything that needs to be exposed externally formalized, and also figure out what the things are that need to be exposed externally.

But unfortunately, I currently can't prioritize it yet.

## Suggested Steps for FastStream

### FastStream's own Dependency Injection system

The first step I would suggest is to not use FastAPI internals, even if you still temporarily use `fastapi.Depends`.

I just merged and released this https://github.com/fastapi/fastapi/pull/14372 for a similar use case that uses `fastapi.Depends` for another tool, but doesn't use and modify other FastAPI internals.

You could start by still supporting users declaring things with `fastapi.Depends`, but solving them in your own way with your own logic.

This first step would probably save you a bunch of headaches related to these changes.

### Deprecation

Checking the tutorial in https://faststream.ag2.ai/latest/getting-started/integrations/fastapi/, I see that there's a "router", nevertheless, I understand it would not be used by FastAPI, it wouldn't handle HTTP requests, it would be used by FastStream to handle event messages, so I don't really see a reason why it would need to be attached to the same FastAPI app.

I would think that logic could (and should) be separated, to separate concerns.

And I would think they should be separate processes, one thing handling the HTTP endpoints and a different one handling stream messages.

From what I understand, the main point of interaction between FastStream and FastAPI is in both using FastAPI's `Depends`.

I would suggest teaching users to separate `fastapi.Depends` and `faststream.Depends`, and in cases where they definitely want to use both for the same type, declare both in the annotation, as `Dep = Annotated[type, fastapi.Depends(get_fastapi_dep), faststream.Depends(get_faststream_dep)`, but I wouldn't even encourage that as I don't think that's necessarily a good idea.

And maybe if you wanted, you could document how to create a FastAPI dependency to get a broker object in a FastAPI endpoint to send messages using FastStream, but none of this would need any plugins for FastAPI in FastStream.

I would probably update that documentation to encourage using the separated libraries. Then I would deprecate the logic in the `faststream.fastapi` plugins, raise warnings when people use it and explain how to migrate to not have both types of apps coupled, and remove documentation teaching the old pattern, and at some point, drop support for (remove) the FastAPI plugins.

I see in that docs page that there's also a bunch of caveats when using the FastAPI plugins, things from FastStream that are not available, and different behaviors. I would try to avoid that as it's more complexity for users to learn, and for you to teach and maintain.

**Comment by Lancetnik** [comment]
Hi @tiangolo 👋
Just a gentle ping to see if you’ve had a chance to look at this yet. Totally understand you’re busy — just wanted to check whether there’s any update or something we could help clarify or contribute to move this forward. Thanks so much for all your work on FastAPI 🙏

**Reply by lfse-slafleur** [reply]
@Lancetnik I might be missing the point here as I found this discussion searching for a solution for a bug. So it feels like answering without fully understanding the discussion from my end. But perhaps, I may offer an alternative. Instead of having a DI system per library that is hard to integrate together, perhaps it might be useful to integrate FastAPI and other libraries into a single, standalone DI system e.g. https://github.com/ets-labs/python-dependency-injector . Not that you really need to do anything to integrate into python-dependency-injector other than provide examples and ensure the examples keep working. But it might make the integration issue listed here obsolete and make it more flexible in the same move.

A change for Faststream would then be to move away from global scoped routers but personally, but that would be (in general) my preference for testability to have as little globally scoped resources as possible. (And I say that without understanding or using Faststream so I might be completely wrong here). Hope this perspective/idea helps!

**Reply by Tishka17** [reply]
From my perspective, dependency-injector is pretty hard to integrate. It has no concept of scopes and resource finalization is done in a very limited way.

From the other hand, to integrate ioc container into a web framework you need a way to tell it to get depdneciea from container and not treat as request data. Also you need a way to get a list of all requested dependencies to validate if they are registered in container. While the first thing can be achieved using additional decorators and request context, the second is hardly achievable in that way, however it won't be perfect from developer experience in that way.

You can check what we've done in `dishka `in terms of integration with different frameworks, it works, but we'd like it to be more convenient
